Sources/CometClient/CometClient.swift

```diff
@@ -0,0 +1,134 @@
+//
+//  CometClient.swift
+//  Comet
+//
+//  Created by Tuan Tu Do on 26.05.2021.
+//  Copyright © 2021 Etnetera. All rights reserved.
+//
+
+import Combine
+import Foundation
+
+// TODO: add refresh token to CometClient
+/// TODO
+/// - Tag: CometClient
+public final class CometClient {
+    private let urlSession: URLSession
+    private let authenticator: Authenticator
+    private let authenticatedRequestBuilder: AuthenticatedRequestBuilding
+    private let requestResponseHandler: RequestResponseHandling
+
+    public init(
+        urlSession: URLSession = .shared,
+        tokenProvider: TokenProviding,
+        authorizedRequestBuilder: AuthenticatedRequestBuilding,
+        requestResponseHandler: RequestResponseHandling = CometRequestResponseHandler()
+    ) {
+        self.urlSession = urlSession
+        self.authenticator = Authenticator(tokenProvider: tokenProvider)
+        self.authenticatedRequestBuilder = authorizedRequestBuilder
+        self.requestResponseHandler = requestResponseHandler
+    }
+
+    public func performAuthenticatedRequest<ResponseObject: Decodable>(
```

Comment:
The `ReponseObject` does not have to implement `Decodable`. For example sometimes you just need to know, that the request was OK and return `Void`.

Reply:
After some thinking I don't think this is a good idea. I think we should stick with enforcing `ResponseObject` to implement `Decodable`. Users will usually use one `RequestResponseHandling` with the `CometClient` and `RequestResponseHandling ` should be generic so it can decode different objects from the data provided by the backend.

Reply:
Instead of returning a ResponseObject, return a Data instead. This allows anyone to use whichever kind of deserialization they desire.

Reply:
Yeah, that's a great idea actually.

Reply:
I will keep this method and add a new one:
``` swift
public func performAuthenticatedRequest(
    _ request: URLRequest
) -> AnyPublisher<(data: Data, response: URLResponse), CometClientError>
```

Reply:
Then you should modify the existing method to take a `TopLevelDecoder` as an additional parameter. Maybe they have a special date format or they want to convert keys from snake case. That also helps if you wanna decode xml or plist or have your own special decoder.

Reply:
I will just keep this way. You can use the method that returns data and do whatever you want to do with it. The existing method will just be used for simple decoding from JSON.

Reply:
And you can actually use your own convertor, if you implement the `RequestResponseHandling`.